pkg/cortex/modules.go

```diff
@@ -506,18 +503,36 @@ func (t *Cortex) initQueryFrontendTripperware() (serv services.Service, err erro
 }
 
 func (t *Cortex) initQueryFrontend() (serv services.Service, err error) {
-	t.Frontend, err = frontend.New(t.Cfg.Frontend, t.Overrides, util.Logger, prometheus.DefaultRegisterer)
+	roundTripper, frontendV1, frontendV2, err := t.Cfg.Frontend.InitFrontend(t.Overrides, t.Cfg.Server.GRPCListenPort, util.Logger, prometheus.DefaultRegisterer)
 	if err != nil {
-		return
+		return nil, err
 	}
 
-	t.Frontend.Wrap(t.QueryFrontendTripperware)
-	t.API.RegisterQueryFrontend(t.Frontend)
+	// Wrap roundtripper into Tripperware.
+	t.QueryFrontendTripperware(roundTripper)
 
-	return services.NewIdleService(nil, func(_ error) error {
-		t.Frontend.Close()
-		return nil
-	}), nil
+	handler := frontend.NewHandler(t.Cfg.Frontend.Handler, roundTripper, util.Logger)
+	if t.Cfg.Frontend.CompressResponses {
+		handler = gziphandler.GzipHandler(handler)
+	}
+
+	t.API.RegisterQueryFrontendHandler(handler)
+
+	if frontendV1 != nil {
```

Comment:
No, that's not possible.

Both query-frontend and querier must scheduler or neither should.

Migration from setup with query-frontend and querier but no scheduler is possible by:

1) starting scheduler (or two)
2) starting query-frontend and querier configured with scheduler
3) redirecting queries from old query-frontend (without scheduler) to new query-frontend
4) once no traffic is coming to old query-frontends, they can be stopped, together with old queriers.

It would be possible with some extra effort to provide migration path at query-frontend level, but this PR doesn't do that.

Reply:
Okay, so basically the migration path would be spin up new queriers + new query frontends + scheduler, then switch traffic from old query frontends and old queriers, then turn down old queriers and old query frontends. That seems like it will work, just requires running more components during transition.